File: solutions/javascript/binary-search-tree/1/binary-search-tree.js
//
// This is only a SKELETON file for the 'Binary Search Tree' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class BinarySearchTree {
  constructor(value) {
    this.value = value;
    this._left = undefined;
    this._right = undefined;
  }

  get data() {
    return this.value;
  }
  get right() {
    return this._right;
  }

  get left() {
    return this._left;
  }

  insert(n) {
    if (n <= this.value) {
      if (this._left) {
        this._left.insert(n);
      } else {
        this._left = new BinarySearchTree(n);
      }
    } else {
      if (this._right) {
        this._right.insert(n);
      } else {
        this._right = new BinarySearchTree(n);
      }
    }
  }

  each(fn) {
    if (this._left) {
      this._left.each(fn);
    }

    fn(this.value);

    if (this._right) {
      this._right.each(fn);
    }
  }

  ceil(n) {
    let ceilValue = undefined;

    function walk(node) {
      if (!node) return;

      if (node.value <= n) {
        walk(node._right);
      } else {
        ceilValue = node.value;
        walk(node._left);
      }
    }

    walk(this);

    return ceilValue;
  }

  floor(n) {
    let floorValue = undefined;

    function walk(node) {
      if (!node) return;

      if (node.value >= n) {
        walk(node._left);
      } else {
        floorValue = node.value;
        walk(node._right);
      }
    }

    walk(this);

    return floorValue;
  }
}

File: solutions/ruby/binary-search-tree/1/binary_search_tree.rb
=begin
Write your code for the 'Binary Search Tree' exercise in this file. Make the tests in
`binary_search_tree_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/binary-search-tree` directory.
=end

class Bst
  attr_reader :data, :left, :right
  
  def initialize(data)
    @data = data
    @left = nil
    @right = nil
  end

  def insert(val)
    if val <= data
      if !left.nil?
        @left.insert(val)
      else
        @left = Bst.new(val)
      end
    else
      if !right.nil?
        @right.insert(val)
      else
        @right = Bst.new(val)
      end
    end
  end

  def each(&block)
    return to_enum(:each) unless block_given?
    
    left&.each(&block)

    yield data

    right&.each(&block)
  end
end